#!/usr/bin/env bash
###############################################################################
# Auto-generated by qiime2 v.2024.10.1 at 10:59:07 AM on 29 Nov, 2024
# This document is a representation of the scholarly work of the creator of the
# QIIME 2 Results provided as input to this software, and may be protected by
# intellectual property law. Please respect all copyright restrictions and
# licenses governing the use, modification, and redistribution of this work.

# For User Support, post to the QIIME2 Forum at https://forum.qiime2.org.

# Instructions for use:
# 1. Open this script in a text editor or IDE. Support for BASH
#    syntax highlighting can be helpful.
# 2. Search or scan visually for '<' or '>' characters to find places where
#    user input (e.g. a filepath or column name) is required. These must be
#    replaced with your own values. E.g. <column name> -> 'patient_id'.
#    Failure to remove '<' or '>' may result in `No such File ...` errors
# 3. Search for 'FIXME' comments in the script, and respond as directed.
# 4. Remove all 'FIXME' comments from the script completely. Failure to do so
#    may result in 'Missing Option' errors
# 5. Adjust the arguments to the commands below to suit your data and metadata.
#    If your data is not identical to that in the replayed analysis,
#    changes may be required. (e.g. sample ids or rarefaction depth)
# 6. Optional: replace any filenames in this script that begin with 'XX' with
#    unique file names to ensure they are preserved. QIIME 2 saves all outputs
#    from all actions in this script to disk regardless of whether those
#    outputs were in the original collection of replayed results. The filenames
#    of "un-replayed" artifacts are prefixed with 'XX' so they may be easily
#    located. These names are not guaranteed to be unique, so 'XX_table.qza'
#    may be overwritten by another 'XX_table.qza' later in the script.
# 7. Activate your replay conda environment, and confirm you have installed all
#    plugins used by the script.
# 8. Run this script with `bash <path to this script>`, or copy-paste commands
#    into the terminal for a more interactive analysis.
# 9. Optional: to delete all results not required to produce the figures and
#    data used to generate this script, navigate to the directory in which you
#    ran the script and `rm XX*.qz*`
###############################################################################
## function to create result collections ##
construct_result_collection () {
	mkdir $rc_name
	touch $rc_name.order
	for key in "${keys[@]}"; do
		echo $key >> $rc_name.order
	done
	for i in "${!keys[@]}"; do
		ln -s ../"${names[i]}" $rc_name"${keys[i]}"$ext
	done
}
##

# This tells bash to -e exit immediately if a command fails
# and -x show all commands in stdout so you can track progress
set -e -x

qiime tools import \
  --type 'NCBIAccessionIDs' \
  --input-path <your data here> \
  --output-path ncbi-accession-i-ds-0.qza

qiime moshpit build-kraken-db \
  --p-collection pluspf \
  --p-threads 1 \
  --p-kmer-len 35 \
  --p-minimizer-len 31 \
  --p-minimizer-spaces 7 \
  --p-no-no-masking \
  --p-max-db-size 0 \
  --p-no-use-ftp \
  --p-load-factor 0.7 \
  --p-no-fast-build \
  --o-kraken2-database kraken2-database-0.qza \
  --o-bracken-database bracken-database-0.qza

qiime moshpit fetch-busco-db \
  --p-virus False \
  --p-prok True \
  --p-euk False \
  --o-busco-db busco-db-0.qza

qiime tools import \
  --type 'FeatureTable[Frequency]' \
  --input-path <your data here> \
  --output-path feature-table-frequency-0.qza

qiime moshpit fetch-kaiju-db \
  --p-database-type nr_euk \
  --o-database database-0.qza

qiime tools import \
  --type 'ReferenceDB[Eggnog]' \
  --input-path <your data here> \
  --output-path reference-db-eggnog-0.qza

qiime tools import \
  --type 'ReferenceDB[Diamond]' \
  --input-path <your data here> \
  --output-path reference-db-diamond-0.qza

qiime fondue get-all \
  --i-accession-ids ncbi-accession-i-ds-0.qza \
  --p-email mziemski@ethz.ch \
  --p-n-jobs 5 \
  --p-retries 5 \
  --p-log-level DEBUG \
  --o-paired-reads paired-reads-0.qza \
  --o-metadata XX_metadata \
  --o-single-reads XX_single_reads \
  --o-failed-runs XX_failed_runs

qiime moshpit classify-kraken2 \
  --i-seqs paired-reads-0.qza \
  --i-kraken2-db kraken2-database-0.qza \
  --p-threads 72 \
  --p-confidence 0.5 \
  --p-minimum-base-quality 0 \
  --p-no-memory-mapping \
  --p-minimum-hit-groups 2 \
  --p-no-quick \
  --p-report-minimizer-data \
  --o-reports reports-0.qza \
  --o-hits XX_hits

qiime assembly assemble-megahit \
  --i-seqs paired-reads-0.qza \
  --p-presets meta-sensitive \
  --p-min-count 2 \
  --p-k-list 21 29 39 59 79 99 119 141 \
  --p-no-no-mercy \
  --p-bubble-level 2 \
  --p-prune-level 2 \
  --p-prune-depth 2 \
  --p-disconnect-ratio 0.1 \
  --p-low-local-ratio 0.2 \
  --p-max-tip-len auto \
  --p-cleaning-rounds 5 \
  --p-no-no-local \
  --p-no-kmin-1pass \
  --p-memory 0.9 \
  --p-mem-flag 1 \
  --p-num-cpu-threads 24 \
  --p-no-no-hw-accel \
  --p-min-contig-len 200 \
  --p-coassemble False \
  --o-contigs contigs-0.qza

qiime moshpit classify-kaiju \
  --i-seqs paired-reads-0.qza \
  --i-db database-0.qza \
  --p-z 16 \
  --p-a greedy \
  --p-e 3 \
  --p-m 11 \
  --p-s 65 \
  --p-evalue 0.01 \
  --p-x \
  --p-r species \
  --p-c 0.1 \
  --p-no-exp \
  --p-no-u \
  --o-taxonomy taxonomy-0.qza \
  --o-abundances XX_abundances

qiime motus profile \
  --i-samples paired-reads-0.qza \
  --p-threads 4 \
  --p-min-alen 75 \
  --p-marker-gene-cutoff 3 \
  --p-mode insert.scaled_counts \
  --p-no-reference-genomes \
  --p-jobs 4 \
  --o-taxonomy taxonomy-1.qza \
  --o-table table-0.qza

qiime moshpit estimate-bracken \
  --i-kraken-reports reports-0.qza \
  --i-bracken-db bracken-database-0.qza \
  --p-threshold 5 \
  --p-read-len 150 \
  --p-level S \
  --o-taxonomy taxonomy-2.qza \
  --o-table table-1.qza \
  --o-reports XX_reports

qiime assembly index-contigs \
  --i-contigs contigs-0.qza \
  --p-no-large-index \
  --p-no-debug \
  --p-no-sanitized \
  --p-verbose \
  --p-no-noauto \
  --p-no-packed \
  --p-bmax auto \
  --p-bmaxdivn 4 \
  --p-dcv 1024 \
  --p-no-nodc \
  --p-offrate 5 \
  --p-ftabchars 10 \
  --p-threads 8 \
  --p-seed 100 \
  --o-index index-0.qza

qiime taxa filter-table \
  --i-table feature-table-frequency-0.qza \
  --i-taxonomy taxonomy-0.qza \
  --p-exclude unclassified,belong,cannot \
  --p-query-delimiter , \
  --p-mode contains \
  --o-filtered-table filtered-table-0.qza

qiime taxa filter-table \
  --i-table table-0.qza \
  --i-taxonomy taxonomy-1.qza \
  --p-exclude 'u; n; a; s; s; i; g; n; e; d' \
  --p-query-delimiter , \
  --p-mode contains \
  --o-filtered-table filtered-table-1.qza

qiime taxa filter-table \
  --i-table table-1.qza \
  --i-taxonomy taxonomy-2.qza \
  --p-exclude Unclassified \
  --p-query-delimiter , \
  --p-mode contains \
  --o-filtered-table filtered-table-2.qza

qiime assembly map-reads \
  --i-index index-0.qza \
  --i-reads paired-reads-0.qza \
  --p-skip 0 \
  --p-qupto unlimited \
  --p-trim5 0 \
  --p-trim3 0 \
  --p-trim-to untrimmed \
  --p-no-phred33 \
  --p-no-phred64 \
  --p-mode local \
  --p-sensitivity sensitive \
  --p-n 0 \
  --p-len 22 \
  --p-i S,1,1.15 \
  --p-n-ceil L,0,0.15 \
  --p-dpad 15 \
  --p-gbar 4 \
  --p-no-ignore-quals \
  --p-no-nofw \
  --p-no-norc \
  --p-no-no-1mm-upfront \
  --p-no-end-to-end \
  --p-no-local \
  --p-ma 2 \
  --p-mp 6 \
  --p-np 1 \
  --p-rdg 5,3 \
  --p-rfg 5,3 \
  --p-k off \
  --p-no-a \
  --p-d 15 \
  --p-r 2 \
  --p-minins 0 \
  --p-maxins 500 \
  --p-valid-mate-orientations fr \
  --p-no-no-mixed \
  --p-no-no-discordant \
  --p-no-dovetail \
  --p-no-no-contain \
  --p-no-no-overlap \
  --p-offrate off \
  --p-threads 12 \
  --p-no-reorder \
  --p-no-mm \
  --p-seed 100 \
  --p-no-non-deterministic \
  --o-alignment-map alignment-map-0.qza

# Replay attempts to represent metadata inputs accurately, but metadata .tsv
# files are merged automatically by some interfaces, rendering distinctions
# between file inputs invisible in provenance. We output the recorded
# metadata to disk to enable visual inspection.

# The following command may have received additional metadata .tsv files. To
# confirm you have covered your metadata needs adequately, review the
# original metadata, saved at '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000g
# q/T/tmpqp6st2e1/provenance_replay/recorded_metadata/diversity_core_metrics
# _0'

qiime diversity core-metrics \
  --i-table filtered-table-0.qza \
  --p-sampling-depth 5298000 \
  --m-metadata-file <your metadata filepath>.tsv \
  --p-no-with-replacement \
  --p-n-jobs 6 \
  --p-no-ignore-missing-samples \
  --o-shannon-vector shannon-vector-0.qza \
  --o-rarefied-table XX_rarefied_table \
  --o-observed-features-vector XX_observed_features_vector \
  --o-evenness-vector XX_evenness_vector \
  --o-jaccard-distance-matrix XX_jaccard_distance_matrix \
  --o-bray-curtis-distance-matrix XX_bray_curtis_distance_matrix \
  --o-jaccard-pcoa-results XX_jaccard_pcoa_results \
  --o-bray-curtis-pcoa-results XX_bray_curtis_pcoa_results \
  --o-jaccard-emperor XX_jaccard_emperor \
  --o-bray-curtis-emperor XX_bray_curtis_emperor

# The following command may have received additional metadata .tsv files. To
# confirm you have covered your metadata needs adequately, review the
# original metadata, saved at '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000g
# q/T/tmpqp6st2e1/provenance_replay/recorded_metadata/diversity_core_metrics
# _1'

qiime diversity core-metrics \
  --i-table filtered-table-1.qza \
  --p-sampling-depth 1792 \
  --m-metadata-file <your metadata filepath>.tsv \
  --p-no-with-replacement \
  --p-n-jobs 6 \
  --p-no-ignore-missing-samples \
  --o-shannon-vector shannon-vector-1.qza \
  --o-rarefied-table XX_rarefied_table \
  --o-observed-features-vector XX_observed_features_vector \
  --o-evenness-vector XX_evenness_vector \
  --o-jaccard-distance-matrix XX_jaccard_distance_matrix \
  --o-bray-curtis-distance-matrix XX_bray_curtis_distance_matrix \
  --o-jaccard-pcoa-results XX_jaccard_pcoa_results \
  --o-bray-curtis-pcoa-results XX_bray_curtis_pcoa_results \
  --o-jaccard-emperor XX_jaccard_emperor \
  --o-bray-curtis-emperor XX_bray_curtis_emperor

# The following command may have received additional metadata .tsv files. To
# confirm you have covered your metadata needs adequately, review the
# original metadata, saved at '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000g
# q/T/tmpqp6st2e1/provenance_replay/recorded_metadata/diversity_core_metrics
# _2'

qiime diversity core-metrics \
  --i-table filtered-table-2.qza \
  --p-sampling-depth 2320000 \
  --m-metadata-file <your metadata filepath>.tsv \
  --p-no-with-replacement \
  --p-n-jobs 6 \
  --p-no-ignore-missing-samples \
  --o-rarefied-table rarefied-table-0.qza \
  --o-shannon-vector shannon-vector-2.qza \
  --o-observed-features-vector XX_observed_features_vector \
  --o-evenness-vector XX_evenness_vector \
  --o-jaccard-distance-matrix XX_jaccard_distance_matrix \
  --o-bray-curtis-distance-matrix XX_bray_curtis_distance_matrix \
  --o-jaccard-pcoa-results XX_jaccard_pcoa_results \
  --o-bray-curtis-pcoa-results XX_bray_curtis_pcoa_results \
  --o-jaccard-emperor XX_jaccard_emperor \
  --o-bray-curtis-emperor XX_bray_curtis_emperor

qiime moshpit bin-contigs-metabat \
  --i-contigs contigs-0.qza \
  --i-alignment-maps alignment-map-0.qza \
  --p-num-threads 64 \
  --p-seed 100 \
  --p-verbose \
  --o-mags mags-0.qza \
  --o-contig-map XX_contig_map \
  --o-unbinned-contigs XX_unbinned_contigs

qiime moshpit evaluate-busco \
  --i-bins mags-0.qza \
  --i-busco-db busco-db-0.qza \
  --p-mode genome \
  --p-lineage-dataset bacteria_odb10 \
  --p-no-augustus \
  --p-auto-lineage False \
  --p-auto-lineage-euk False \
  --p-auto-lineage-prok False \
  --p-cpu 16 \
  --p-contig-break 10 \
  --p-evalue 0.001 \
  --p-no-force \
  --p-limit 3 \
  --p-no-long \
  --p-no-miniprot \
  --p-no-scaffold-composition \
  --o-results-table results-table-0.qza \
  --o-visualization XX_visualization

# The following command may have received additional metadata .tsv files. To
# confirm you have covered your metadata needs adequately, review the
# original metadata, saved at '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000g
# q/T/tmpqp6st2e1/provenance_replay/recorded_metadata/moshpit_filter_mags_0'

qiime moshpit filter-mags \
  --i-mags mags-0.qza \
  --m-metadata-file results-table-0.qza \
  --p-where 'complete>50' \
  --p-no-exclude-ids \
  --p-on mag \
  --o-filtered-mags filtered-mags-0.qza

# The following command may have received additional metadata .tsv files. To
# confirm you have covered your metadata needs adequately, review the
# original metadata, saved at '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000g
# q/T/tmpqp6st2e1/provenance_replay/recorded_metadata/moshpit_filter_mags_1'

qiime moshpit filter-mags \
  --i-mags mags-0.qza \
  --m-metadata-file results-table-0.qza \
  --p-where 'complete>90' \
  --p-no-exclude-ids \
  --p-on mag \
  --o-filtered-mags filtered-mags-1.qza

qiime sourmash compute \
  --i-sequence-file filtered-mags-0.qza \
  --p-ksizes 35 \
  --p-scaled 10 \
  --p-track-abundance \
  --o-min-hash-signature min-hash-signature-0.qza

qiime sourmash compute \
  --i-sequence-file filtered-mags-1.qza \
  --p-ksizes 35 \
  --p-scaled 10 \
  --p-track-abundance \
  --o-min-hash-signature min-hash-signature-1.qza

qiime sourmash compare \
  --i-min-hash-signature min-hash-signature-0.qza \
  --p-ksize 35 \
  --p-ignore-abundance \
  --o-compare-output compare-output-0.qza

qiime sourmash compare \
  --i-min-hash-signature min-hash-signature-1.qza \
  --p-ksize 35 \
  --p-ignore-abundance \
  --o-compare-output compare-output-1.qza

qiime moshpit dereplicate-mags \
  --i-mags filtered-mags-0.qza \
  --i-distance-matrix compare-output-0.qza \
  --p-threshold 0.99 \
  --o-dereplicated-mags dereplicated-mags-0.qza \
  --o-feature-table XX_feature_table

qiime moshpit dereplicate-mags \
  --i-mags filtered-mags-1.qza \
  --i-distance-matrix compare-output-1.qza \
  --p-threshold 0.99 \
  --o-dereplicated-mags dereplicated-mags-1.qza \
  --o-feature-table XX_feature_table

qiime moshpit get-feature-lengths \
  --i-features dereplicated-mags-0.qza \
  --o-lengths lengths-0.qza

qiime assembly index-derep-mags \
  --i-mags dereplicated-mags-0.qza \
  --p-no-large-index \
  --p-no-debug \
  --p-no-sanitized \
  --p-verbose \
  --p-no-noauto \
  --p-no-packed \
  --p-bmax auto \
  --p-bmaxdivn 4 \
  --p-dcv 1024 \
  --p-no-nodc \
  --p-offrate 5 \
  --p-ftabchars 10 \
  --p-threads 8 \
  --p-seed 100 \
  --o-index index-1.qza

qiime moshpit classify-kraken2 \
  --i-seqs dereplicated-mags-0.qza \
  --i-kraken2-db kraken2-database-0.qza \
  --p-threads 72 \
  --p-confidence 0.5 \
  --p-minimum-base-quality 0 \
  --p-no-memory-mapping \
  --p-minimum-hit-groups 2 \
  --p-no-quick \
  --p-report-minimizer-data \
  --o-hits hits-0.qza \
  --o-reports reports-1.qza

qiime assembly index-derep-mags \
  --i-mags dereplicated-mags-1.qza \
  --p-no-large-index \
  --p-no-debug \
  --p-no-sanitized \
  --p-verbose \
  --p-no-noauto \
  --p-no-packed \
  --p-bmax auto \
  --p-bmaxdivn 4 \
  --p-dcv 1024 \
  --p-no-nodc \
  --p-offrate 5 \
  --p-ftabchars 10 \
  --p-threads 8 \
  --p-seed 100 \
  --o-index index-2.qza

qiime moshpit eggnog-diamond-search \
  --i-sequences dereplicated-mags-1.qza \
  --i-diamond-db reference-db-diamond-0.qza \
  --p-num-cpus 16 \
  --p-db-in-memory \
  --o-eggnog-hits eggnog-hits-0.qza \
  --o-table XX_table

qiime moshpit get-feature-lengths \
  --i-features dereplicated-mags-1.qza \
  --o-lengths lengths-1.qza

qiime assembly map-reads \
  --i-index index-1.qza \
  --i-reads paired-reads-0.qza \
  --p-skip 0 \
  --p-qupto unlimited \
  --p-trim5 0 \
  --p-trim3 0 \
  --p-trim-to untrimmed \
  --p-no-phred33 \
  --p-no-phred64 \
  --p-mode local \
  --p-sensitivity sensitive \
  --p-n 0 \
  --p-len 22 \
  --p-i S,1,1.15 \
  --p-n-ceil L,0,0.15 \
  --p-dpad 15 \
  --p-gbar 4 \
  --p-no-ignore-quals \
  --p-no-nofw \
  --p-no-norc \
  --p-no-no-1mm-upfront \
  --p-no-end-to-end \
  --p-no-local \
  --p-ma 2 \
  --p-mp 6 \
  --p-np 1 \
  --p-rdg 5,3 \
  --p-rfg 5,3 \
  --p-k off \
  --p-no-a \
  --p-d 15 \
  --p-r 2 \
  --p-minins 0 \
  --p-maxins 500 \
  --p-valid-mate-orientations fr \
  --p-no-no-mixed \
  --p-no-no-discordant \
  --p-no-dovetail \
  --p-no-no-contain \
  --p-no-no-overlap \
  --p-offrate off \
  --p-threads 12 \
  --p-no-reorder \
  --p-no-mm \
  --p-seed 100 \
  --p-no-non-deterministic \
  --o-alignment-map alignment-map-1.qza

qiime moshpit kraken2-to-mag-features \
  --i-reports reports-1.qza \
  --i-hits hits-0.qza \
  --p-coverage-threshold 0.1 \
  --o-taxonomy taxonomy-3.qza

qiime assembly map-reads \
  --i-index index-2.qza \
  --i-reads paired-reads-0.qza \
  --p-skip 0 \
  --p-qupto unlimited \
  --p-trim5 0 \
  --p-trim3 0 \
  --p-trim-to untrimmed \
  --p-no-phred33 \
  --p-no-phred64 \
  --p-mode local \
  --p-sensitivity sensitive \
  --p-n 0 \
  --p-len 22 \
  --p-i S,1,1.15 \
  --p-n-ceil L,0,0.15 \
  --p-dpad 15 \
  --p-gbar 4 \
  --p-no-ignore-quals \
  --p-no-nofw \
  --p-no-norc \
  --p-no-no-1mm-upfront \
  --p-no-end-to-end \
  --p-no-local \
  --p-ma 2 \
  --p-mp 6 \
  --p-np 1 \
  --p-rdg 5,3 \
  --p-rfg 5,3 \
  --p-k off \
  --p-no-a \
  --p-d 15 \
  --p-r 2 \
  --p-minins 0 \
  --p-maxins 500 \
  --p-valid-mate-orientations fr \
  --p-no-no-mixed \
  --p-no-no-discordant \
  --p-no-dovetail \
  --p-no-no-contain \
  --p-no-no-overlap \
  --p-offrate off \
  --p-threads 12 \
  --p-no-reorder \
  --p-no-mm \
  --p-seed 100 \
  --p-no-non-deterministic \
  --o-alignment-map alignment-map-2.qza

qiime moshpit eggnog-annotate \
  --i-eggnog-hits eggnog-hits-0.qza \
  --i-eggnog-db reference-db-eggnog-0.qza \
  --p-db-in-memory \
  --p-num-cpus 16 \
  --o-ortholog-annotations ortholog-annotations-0.qza

qiime moshpit estimate-mag-abundance \
  --i-maps alignment-map-1.qza \
  --i-mag-lengths lengths-0.qza \
  --p-metric rpkm \
  --p-min-mapq 42 \
  --p-min-query-len 0 \
  --p-min-base-quality 20 \
  --p-min-read-len 0 \
  --p-threads 24 \
  --o-abundances abundances-0.qza

qiime moshpit estimate-mag-abundance \
  --i-maps alignment-map-2.qza \
  --i-mag-lengths lengths-1.qza \
  --p-metric rpkm \
  --p-min-mapq 42 \
  --p-min-query-len 0 \
  --p-min-base-quality 20 \
  --p-min-read-len 0 \
  --p-threads 24 \
  --o-abundances abundances-1.qza

qiime moshpit extract-annotations \
  --i-ortholog-annotations ortholog-annotations-0.qza \
  --p-annotation caz \
  --p-max-evalue 0.0001 \
  --p-min-score 0.0 \
  --o-annotation-frequency annotation-frequency-0.qza

# The following command may have received additional metadata .tsv files. To
# confirm you have covered your metadata needs adequately, review the
# original metadata, saved at '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000g
# q/T/tmpqp6st2e1/provenance_replay/recorded_metadata/diversity_core_metrics
# _3'

qiime diversity core-metrics \
  --i-table abundances-0.qza \
  --p-sampling-depth 210 \
  --m-metadata-file <your metadata filepath>.tsv \
  --p-no-with-replacement \
  --p-n-jobs 6 \
  --p-no-ignore-missing-samples \
  --o-shannon-vector shannon-vector-3.qza \
  --o-rarefied-table rarefied-table-1.qza \
  --o-observed-features-vector XX_observed_features_vector \
  --o-evenness-vector XX_evenness_vector \
  --o-jaccard-distance-matrix XX_jaccard_distance_matrix \
  --o-bray-curtis-distance-matrix XX_bray_curtis_distance_matrix \
  --o-jaccard-pcoa-results XX_jaccard_pcoa_results \
  --o-bray-curtis-pcoa-results XX_bray_curtis_pcoa_results \
  --o-jaccard-emperor XX_jaccard_emperor \
  --o-bray-curtis-emperor XX_bray_curtis_emperor

qiime moshpit multiply-tables \
  --i-table1 abundances-1.qza \
  --i-table2 annotation-frequency-0.qza \
  --o-result-table result-table-0.qza

# The following command may have received additional metadata .tsv files. To
# confirm you have covered your metadata needs adequately, review the
# original metadata, saved at '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000g
# q/T/tmpqp6st2e1/provenance_replay/recorded_metadata/diversity_core_metrics
# _4'

qiime diversity core-metrics \
  --i-table result-table-0.qza \
  --p-sampling-depth 8000 \
  --m-metadata-file <your metadata filepath>.tsv \
  --p-no-with-replacement \
  --p-n-jobs 6 \
  --p-no-ignore-missing-samples \
  --o-rarefied-table rarefied-table-2.qza \
  --o-shannon-vector shannon-vector-4.qza \
  --o-observed-features-vector XX_observed_features_vector \
  --o-evenness-vector XX_evenness_vector \
  --o-jaccard-distance-matrix XX_jaccard_distance_matrix \
  --o-bray-curtis-distance-matrix XX_bray_curtis_distance_matrix \
  --o-jaccard-pcoa-results XX_jaccard_pcoa_results \
  --o-bray-curtis-pcoa-results XX_bray_curtis_pcoa_results \
  --o-jaccard-emperor XX_jaccard_emperor \
  --o-bray-curtis-emperor XX_bray_curtis_emperor


###############################################################################
# The following QIIME 2 Results were parsed to produce this script:
# 137046cb-fe18-4d0b-babc-de209fb49e65 	 2b5122a9-0396-4ab7-966b-f66fa8a00705
# 5a0bf237-2368-4fe7-9901-50d9d6d92dfb 	 6cc70426-844c-47ec-82ee-9be5a2d90cda
# 7f0e9ecb-01f7-4a4c-b38f-0bee505d5f76 	 af746e01-9abf-4d0f-89bb-8abe68b38705
# b6ac8052-ee05-4029-b871-6b2202641a86 	 bf3224d3-1051-47f4-b79f-c1997f22e1b6
# ca4c6fe6-66a7-41ca-b434-bab6f1cc6d18 	 e2c284d9-3b8b-4de8-8c19-dc090a0a07e0
###############################################################################
